### PR TITLE
Create custom theme container for epub renderer [GSoC '23]

### PR DESCRIPTION
**Pull request by akila-i**
## Summary

> - Kolibri EPUB renderer currently has 6 fixed built in themes, which are not allowed to be edited by users. Also, Kolibri does not allow adding new color themes to the EPUB renderer with the colors of user's preference. Changes in this PR will bring the ability for the users to add, edit and delete custom color themes for the epub viewer as they prefer.
> - User Interfaces : 
- - **Before** - 
image

- - **Current status** - 
1.1 Settings side bar in EPUB viewer (Desktop)
<p>
image
</p>
1.2 Settings side bar in EPUB viewer (Mobile)
<p>
image
</p>
2.1 Add a theme modal (Desktop)
<p>
image
</p>
2.2 Add a theme modal (Mobile)
<p>
image
</p>
3.1 Select a color using the color picker (Desktop) [CSS to be added]
<p>
image
</p>
3.2 Select a color using the color picker (Mobile)  [CSS to be added]
<p>
image
</p>

## References

> - #10741 
> - Color picker used : [vue-color](https://github.com/xiaokaike/vue-color)

## Reviewer guidance

> Can try the following features
> - Adding a new theme by selecting preferred colors for the text and background (link color is not implemented yet)
> - Deleting a custom created theme
> - Applying a custom created theme to the EPUB viewer
> - Editing a custom created theme to change the colors (right now it creates a new theme with the changed colors as the themeName property is not implemented yet)

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

File: kolibri/plugins/epub_viewer/assets/src/views/EpubConstants.js [hunk 1]
```diff
@@ -40,3 +40,49 @@ export const THEMES = {
 };
 
 export const darkThemeNames = ['BLACK', 'GREY', 'YELLOW'];
+
+// sample custom themes for reference
```
Comment (akila-i) [inline kolibri/plugins/epub_viewer/assets/src/views/EpubConstants.js]: These are no longer needed and will be removed in an upcoming commit

File: kolibri/plugins/epub_viewer/assets/src/views/ThemeNameTextBox.vue [hunk 2]
```diff
@@ -0,0 +1,80 @@
+<template>
+
+  <KTextbox
```
Comment (akila-i) [inline kolibri/plugins/epub_viewer/assets/src/views/ThemeNameTextBox.vue]: This component is yet to be implemented

File: yarn.lock
Comment (akila-i) [inline yarn.lock:1]: Does this file supposed to give a huge diff like this when a new dependancy is added?

Comment (rtibbles) [reply yarn.lock:1]: It can happen - but this does seem unusually large.

Comment (akila-i) [reply yarn.lock:1]: reverted back the yarn.lock file for now. Will try yarn install on a fresh venv, observe the changes and push the yarn.lock file in an upcoming commit.

**Comment by akila-i** [comment]
Linters need to be run and the changes need to be pushed (**TODO**)

**Comment by akila-i** [comment]
> Linters need to be run and the changes need to be pushed (**TODO**)

Linting test failures are fixed.

File: kolibri/plugins/epub_viewer/assets/src/views/EpubConstants.js (hunk 1 above)
Comment (akila-i) [reply kolibri/plugins/epub_viewer/assets/src/views/EpubConstants.js]: > These are no longer needed and will be removed in an upcoming commit

Removed

File: yarn.lock
Comment (akila-i) [reply yarn.lock:1]: added yarn.lock file back with only the expected changes

**Comment by Akila-I** [comment]
> Linters need to be run and the changes need to be pushed (**TODO**)

Done

File: kolibri/plugins/epub_viewer/assets/src/views/ThemeNameTextBox.vue (hunk 2 above)
Comment (akila-i) [reply kolibri/plugins/epub_viewer/assets/src/views/ThemeNameTextBox.vue]: Done

**Comment by tomiwaoLE** [comment]
@marcellamaki just wondering if I can be any help here?

**Comment by marcellamaki** [comment]
@tomiwaoLE - yes but probably in the follow up PR. This is just the MVP for functionality, and @Akila-I has some other in-progress work that includes the designs that you worked on, which would be great to have your feedback on. I'll be sure tag you for review when that's ready :) thank you!!

**Comment by radinamatic** [comment]
Hey @Akila-I ! I'm still trying to finish the full test with the screen reader, but there is a one set of issues that you can already jump on fixing: the 3 new buttons that are in the `New/Edit theme` dialog (`Background`, `Text`, `Link`) do have visible labels, but those are not properly associated with the respective buttons:

Win10-2304 (Akila)  Running  - Oracle VM VirtualBox_001

See how in the NVDA Speech log it can only announce **'button'** and does not say the labels? Compare with how the labels for '`Cancel`' and '`Save`' are properly announced at the end:


https://github.com/learningequality/kolibri/assets/1457929/5a83acf8-c711-4d10-a1a5-ec24e7711c7e

**Comment by akila-i** [comment]
> Hey @Akila-I ! I'm still trying to finish the full test with the screen reader, but there is a one set of issues that you can already jump on fixing: the 3 new buttons that are in the `New/Edit theme` dialog (`Background`, `Text`, `Link`) do have visible labels, but those are not properly associated with the respective buttons:
> 
> Win10-2304 (Akila) Running - Oracle VM VirtualBox_001
> 
> See how in the NVDA Speech log it can only announce **'button'** and does not say the labels? Compare with how the labels for '`Cancel`' and '`Save`' are properly announced at the end:
> 
>  labels.mp4

Thanks for pointing it out. Will fix that before the sync up

**Comment by akila-i** [comment]
> > Hey @Akila-I ! I'm still trying to finish the full test with the screen reader, but there is a one set of issues that you can already jump on fixing: the 3 new buttons that are in the `New/Edit theme` dialog (`Background`, `Text`, `Link`) do have visible labels, but those are not properly associated with the respective buttons:
> > Win10-2304 (Akila) Running - Oracle VM VirtualBox_001
> > See how in the NVDA Speech log it can only announce **'button'** and does not say the labels? Compare with how the labels for '`Cancel`' and '`Save`' are properly announced at the end:
> > labels.mp4
> 
> Thanks for pointing it out. Will fix that before the sync up

H @radinamatic , this issue has been fixed. Please test and confirm

**Comment by marcellamaki** [comment]
Closing as this work has been superseded by the work completed by @Akila-I and merged into `[epub-custom-themes](https://github.com/learningequality/kolibri/tree/epub-custom-themes)` (#11156 and https://github.com/learningequality/kolibri/pull/11041)

**Pull request closed without merging**